File: crates/wormhole-redirector/src/redirector.rs
use crate::Result;
use async_trait::async_trait;
use wormhole_core::{ShortCode, UrlRecord};

#[async_trait]
pub trait Redirector: Send + Sync + 'static {
    /// Resolves a short code to its stored URL record.
    /// Returns `None` if the code does not exist or has expired.
    async fn resolve(&self, code: &ShortCode) -> Result<Option<UrlRecord>>;
}

File: services/redirector/internal/service/redirector.go
// Package service defines the RedirectorService interface.
package service

import (
	"context"

	"github.com/Devin-Yeung/wormhole/services/redirector/internal/domain"
)

// RedirectorService is the primary port for URL redirection operations.
type RedirectorService interface {
	// Resolve returns the UrlRecord for the given short code.
	// Returns nil without error if the code does not exist.
	Resolve(ctx context.Context, shortCode string) (*domain.UrlRecord, error)
}

// TODO: implement RedirectorServiceImpl backed by repository + cache
